segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadOperation.java

```diff
@@ -141,23 +143,56 @@ private void logEnd() {
                     } else {
                         Preconditions.checkState(bytesToRead != 0, "bytesToRead is 0");
                         // Read data from the chunk.
-                        return chunkedSegmentStorage.getChunkStorage().openRead(chunkToReadFrom.getName())
-                                .thenComposeAsync(chunkHandle ->
-                                        chunkedSegmentStorage.getChunkStorage().read(chunkHandle,
-                                                currentOffset.get() - startOffsetForCurrentChunk.get(),
-                                                bytesToRead,
-                                                buffer,
-                                                currentBufferOffset.get())
-                                                .thenAcceptAsync(bytesRead -> {
-                                                    bytesRemaining.addAndGet(-bytesRead);
-                                                    currentOffset.addAndGet(bytesRead);
-                                                    currentBufferOffset.addAndGet(bytesRead);
-                                                    totalBytesRead.addAndGet(bytesRead);
-                                                }, chunkedSegmentStorage.getExecutor()),
-                                        chunkedSegmentStorage.getExecutor());
+                        return CompletableFuture.runAsync(() -> {
+                            // Create parallel requests to read each chunk.
+                            chunkReadFutures.add(readChunk(chunkToReadFrom.getName(),
+                                    currentOffset.get() - startOffsetForCurrentChunk.get(),
+                                    bytesToRead,
+                                    currentBufferOffset.get()));
+                            log.trace("{} read - reading chunk - op={}, segment={}, chunk={} offset={} length={} bufferOffset={}",
+                                    chunkedSegmentStorage.getLogPrefix(),
+                                    System.identityHashCode(this), handle.getSegmentName(),
+                                    chunkToReadFrom.getName(),
+                                    currentOffset.get() - startOffsetForCurrentChunk.get(),
+                                    bytesToRead,
+                                    currentBufferOffset.get());
+                            bytesRemaining.addAndGet(-bytesToRead);
```

Comment:
The previous logic only modified all these counters when the read request completed. Instead, your are modifying them immediately (i.e., the process waits for all the requests to complete at then, but these atomic variables would have been already incremented/decremented). This is a change in behavior that may have side effects for the `ReadOperation` state, but I don't see any test verifying that the result is still valid.

Reply:
The counters are used only to calculate what chunk to read and at what offset in byte array to copy data. This change does not change that logic. It only parallelizes actual reading of individual chunks.